src/polygon/src/polygon.cpp

```diff
@@ -265,34 +265,53 @@ class EarClip {
     // it finds a clear geometric result. In the vast majority of cases the loop
     // will never run, and when it does, it usually only needs one iteration.
     bool IsConvex(float precision) const {
-      int convexity = CCW(left->pos, pos, right->pos, precision);
-      if (convexity != 0) {
-        return convexity > 0;
-      }
-
-      // Uncertain - walk the polygon to get certainty.
+      const float p2 = precision * precision;
```

Comment:
The second test case problem came from failing to clip degenerate triangles, the reason being that IsConvex had a significant bug. I ended up going back to my old IsHole function and fixing it up for this, since I'd done a better job with it before.